download/download.go

```diff
@@ -4,150 +4,204 @@ import (
+	}
+
+	var urls []string
+	d.Find("a.external-link").Each(func(_ int, a *goquery.Selection) {
+		h, exist := a.Attr("href")
+		if !exist {
+			return
+		}
+		if strings.HasSuffix(h, ".zip") {
+			urls = append(urls, h)
+		}
+	})
+	return urls, nil
 }
 
-func getFiles(m bool, dir string) []file {
+func getFiles(dir string) ([]file, error) {
 	fs := []file{{
-		url:   "https://cnae.ibge.gov.br/images/concla/documentacao/CNAE_Subclasses_2_3_Estrutura_Detalhada.xlsx",
-		path:  filepath.Join(dir, "CNAE_Subclasses_2_3_Estrutura_Detalhada.xlsx"),
-		extra: true,
+		url:  listOfCNAE,
+		path: filepath.Join(dir, "CNAE_Subclasses_2_3_Estrutura_Detalhada.xlsx"),
 	}}
 
-	var s string
-	if m {
-		s = brasilIO
-	} else {
-		s = federalRevenue
+	urls, err := getURLs(federalRevenue)
+	if err != nil {
+		return fs, err
 	}
-	for i := 1; i <= files; i++ {
-		n := fmt.Sprintf(FilePattern, i)
-		fs = append(fs, file{url: fmt.Sprintf("%s%s", s, n), path: filepath.Join(dir, n)})
+
+	for _, u := range urls {
+		fs = append(fs, file{url: u, path: filepath.Join(dir, u[strings.LastIndex(u, "/")+1:])})
 	}
-	return fs
+	return fs, nil
+}
+
+type Downloader struct {
+	files     []file
+	client    *http.Client
+	totalSize int64
+	bar       *progressbar.ProgressBar
 }
 
-func getSize(c chan<- size, url string) {
-	var size size
-	var r *http.Response
-	r, size.err = http.Head(url)
-	if size.err != nil {
-		c <- size
+func (d *Downloader) getSize(ch chan<- error, url string) {
+	// IBGE does not respond to HTTP HEAD as expected
+	if url == listOfCNAE {
+		d.totalSize += 137216
+		ch <- nil
 		return
 	}
 
+	r, err := d.client.Head(url)
+	if err != nil {
+		ch <- err
+		return
+	}
 	for _, k := range []string{"Content-Length", "content-length"} {
-		size.size, size.err = strconv.Atoi(r.Header.Get(k))
-		if size.err == nil {
-			c <- size
+		var s int
+		s, err = strconv.Atoi(r.Header.Get(k))
+		if err == nil {
+			d.totalSize += int64(s)
+			ch <- nil
 			return
 		}
 	}
+	err = fmt.Errorf("Could not get size for %s", url)
+	ch <- err
+}
+
+func (d *Downloader) getTotalSize() error {
+	d.totalSize = 0
+	q := make(chan error)
+	for _, f := range d.files {
+		go d.getSize(q, f.url)
+	}
+	for range d.files {
+		err := <-q
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
 
-	size.err = fmt.Errorf("Could not get size for %s", url)
-	c <- size
-	return
+func (d *Downloader) setProgressBar() {
+	d.bar = progressbar.DefaultBytes(d.totalSize, "Downloading")
 }
 
-func download(c chan<- error, b *progressbar.ProgressBar, f file) {
-	r, err := http.Get(f.url)
+func (d *Downloader) download(ch chan<- error, f *file, a int) {
+	r, err := d.client.Get(f.url)
 	if err != nil {
-		c <- err
+		log.Output(2, fmt.Sprintf("HTTP request to %s failed", f.url))
+		ch <- err
 		return
 	}
 	defer r.Body.Close()
 
-	h, err := os.Create(f.path)
+	if r.StatusCode != http.StatusOK {
+		if a < retries {
+			time.Sleep(time.Duration(int(math.Pow(float64(2), float64(a)))) * time.Second)
```

Comment:
Exponential backoff FTW!!!!

Reply:
You love recursive functions
I love exponential backoffs 
What a pair : )